File: CodingInterviewsPackSolutionsCPP/951. Flip Equivalent Binary Trees.cpp
class Solution {
public:
    string toStr(int n){
        ostringstream oss;
        oss<<n;
        return oss.str();
    }
    
    string parenthesize_canonical(TreeNode* root){
        if(!root)
            return "()";
        
        string left = parenthesize_canonical(root->left);
        string right = parenthesize_canonical(root->right);
        
        if(left>right)
            swap(left,right);
        
        return "("+toStr(root->val)+left+right+")";
    }
    
    bool flipEquiv(TreeNode* root1, TreeNode* root2) {
        return parenthesize_canonical(root1)==parenthesize_canonical(root2);
    }
};

File: CodingInterviewsPythonSolutions/951. Flip Equivalent Binary Trees.py
class Solution:
    def toStr(self, n):
        return str(n)
    
    def parenthesize_canonical(self, root):
        if not root:
            return "()"
        
        left = self.parenthesize_canonical(root.left)
        right = self.parenthesize_canonical(root.right)
        
        if left > right:
            left, right = right, left
        
        return "(" + self.toStr(root.val) + left + right + ")"
    
    def flipEquiv(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return self.parenthesize_canonical(root1) == self.parenthesize_canonical(root2)
